### ISSUE TITLE
[tlul] How PutFullData vs PutPartialData should be used

### ISSUE THREAD

**Issue by GregAC** [lowRISC/opentitan#1867]
Tilelink has two separate write messages `PutFullData` and `PutPartialData`. When `PutFullData` is used every byte must be written as indicated by the `a_mask` field of the message payload, where the written bytes are a function of `a_size` and `a_address`.

For Ibex and the RISC-V debug module this presents a slight difficulty, they give no indication as to whether their write is full or partial. However this can be determined by looking at the byte-enables. The question is whether we want to bother with this logic?

It is perfectly permissible for `PutPartialData` to write every byte, `PutFullData` is really only an optimisation which is most useful in the TL-UH variation of the protocol which includes bursts. Here `PutFullData` allows an agent to know every byte is being written before it has seen every message of the burst. We don't have bursts in TL-UL so we don't have this reason to use `PutFullData`

On the other hand some agents may only wish to implement `PutFullData`, for example a device may decide it only wants full 32-bit accesses, from my reading of the Tilelink specification it would be permissible for such a device to return an error on any `PutPartialData` message or `PutFullData` message with a non 32-bit size.

One path forward would be to say the Ibex adapter (to keep it as simple as possible) only emits `PutPartialData` in which cases devices must be able to to deal with `PutPartialData`. Another possibility would be to alter the Ibex top-level interface to make it simple to choose between `PutFullData` and `PutPartialData` (see https://github.com/lowRISC/ibex/issues/757).

In conclusion I think we need to decide

1. Do we want the Ibex TL-UL adapater to emit `PutFullData` when it's a full data write or are we happy with `PutPartialData` always?

2. Can devices indicate an error response on sizes/masks they don't like? Must they have to handle `PutPartialData` or could they always indicate error for `PutPartialData` (this is effectively decision 1 in reverse).

We should add something to the Comportability guidelines at the conclusion of this discussions.

**Comment by GregAC** [comment]
@msfschaffner @eunchan @weicaiyang @sjgitty any thoughts?

**Comment by imphil** [comment]
The basic assumption for all TL-UL in OT is that `size` is always a full 32 bit word.

Am I correct that the underlying motivation for this discussion is that this assumption is too strict and should change?

For (1): we already implement that logic in https://github.com/lowRISC/opentitan/blob/bce0b73e870f91e3a0fb2f9b12fc302523f161c7/hw/ip/rv_core_ibex/rtl/rv_core_ibex.sv#L245-L246.

**Comment by GregAC** [comment]
This may be fine but we should ensure it's documented in the Comportability specifications.

Just realised we have our own TL-UL spec: https://docs.opentitan.org/hw/ip/tlul/doc/ which does basically answer this issue.

It does state that register accesses must be 32-bit aligned which effectively implies 32-bit size always is fine. Though I think we should make this requirement more prominent (as it it's one point of many of the various things the protocol checker enforces)

Also says it's fine for error responses on non-supported accesses.

We still have the question of whether we're happy with Ibex adapter emitting `PutPartialData` only but if we're mandating all device accesses are 32-bit aligned 32-bit size then generating `PutFullData` vs `PutPartialData` is trivial anyway.

So I think answers to my questions would be:

1. Ibex adapter should emit PutFullData where appropriate
2. Devices can give errors on accesses they don't like, permissible for it to only accept `PutFullData`

**Comment by GregAC** [comment]
Also some more detail in the register tool documentation: https://docs.opentitan.org/doc/rm/register_tool/#error-responses

Not 100% certain a device would be free to signal error (or the security form of error which is random data response or silent no-write) on a `PutPartialData` it chose to but it doesn't seem unreasonable given the other rules/requirements. Would save devices having to check for full byte-enables on a `PutPartialData`.

**Comment by imphil** [comment]
Yes, devices can error on PutPartialData requests. (The wording leave a bit room for interpretation either way, but that's what I remember from back then.)

https://docs.opentitan.org/hw/ip/tlul/doc/, "Usage of Opcode, Size and Mask": "Each bus device has an option to support or not support the full variety of bus transaction sizes. Their support will be documented in the device specification."

Maybe you can improve the docs to make these points clearer, and potentially add assertions into the [TL-UL protocol checker](https://docs.opentitan.org/hw/ip/tlul/doc/TlulProtocolChecker/) if they aren't covering all cases at this point? (That's where the assertions I thought about went.)

**Comment by GregAC** [comment]
Yeah I think the ultimate action from this issue is re-arranging the docs a bit, in particular I don't think we should write about the requirements within the register_tool documentation. I will do a PR with a suggested re-arrangement.

**Comment by GregAC** [comment]
One implication of the current setup means an `lb` or `lh` instruction targeted at a device address will generate a 32-bit read at that device, Ibex will then select out the appropriate byte/half-word and place it in the destination register without any error. This may not be a big issue but the converse an `sb` or `sh` targeted at the same address could result in an error response (as it'll be `PutPartialData` without a full mask which the device can respond with an error too).

This could be a surprising behavior to a programmer reading the documentation and seeing only 32-bit aligned accesses to a device are allowed?

**Comment by GregAC** [comment]
Another more problematic behaviour would be an unaligned lw to a device address will end up generating two accesses, both of which will complete without a problem. Again a programmer may expect an exception here.

As this has become a bit rambling and could use some SW input I will start another issue...

**Comment by imphil** [comment]
Yeah, I think software tripped over that in the past. As path forward, let's

1. Fix RTL to close #53 and #1860, keeping the current assumption of size == 32 bit word.
2. Update the docs to make that assumption clearer.
3. Discuss if this assumption is actually a good one, esp. regarding SW ergonomics. This should at least result in a bit of docs better explaining whatever behavior we end up using.

**Comment by asb** [comment]
I think that's a great path forwards for clarifying requirements on the OpenTitan side. After doing so, we probably want to have a separate discussion (hinted at by Greg above) about whether the limitation of only supporting 32-bit device accesses is reasonable for Ibex. In the absence of available TL-UL IP that we want to interface with that does support byte/half-word writes, I suspect it's a documented limitation that we'd be open to removing in the future (e.g. if someone has IP that supports smaller access widths).

**Comment by msfschaffner** [comment]
@GregAC can this issue be updated to reflect the current state?

**Comment by GregAC** [comment]
No design changes required here, one for https://github.com/lowRISC/opentitan/labels/Type%3AFutureRelease I think as it may become more relevant when we have a richer OT ecosystem with more top levels and other IPs and protocol issues like this should be properly resolved.

With our current setup I'm confident things are consistent so there's no issue to worry about for M2.5.

**Comment by vogelpi** [comment]
Triaged for Earl Grey 2. It would be good to document the current use of this field and error responses. But it's not a priority. Another issue in the same corner (document how we implement the spec) is this one here: #30952.
